### PR TITLE
Add simple Catboost model

### PR DESCRIPTION
**Pull request by yonromai**
Adds a trivial use of the Catboost model to predict the `rs_classification` label (from `efo_otar_slim_v3.43.0_rs_classification.tsv`).

Note: The code isn't intended to be especially useful as of now. Rather, it is meant to serve as a first baseline & anchor for future feature and model developments. 

---
Example of run:

```shell
Learning rate set to 0.091517
0:      learn: 1.2655947        total: 74.3ms   remaining: 1m 14s
100:    learn: 0.4192182        total: 1.04s    remaining: 9.3s
200:    learn: 0.3914375        total: 2.02s    remaining: 8.03s
300:    learn: 0.3724496        total: 3s       remaining: 6.98s
400:    learn: 0.3568311        total: 4.1s     remaining: 6.13s
500:    learn: 0.3437438        total: 5.1s     remaining: 5.08s
600:    learn: 0.3304904        total: 6.09s    remaining: 4.04s
700:    learn: 0.3194607        total: 7.26s    remaining: 3.1s
800:    learn: 0.3095234        total: 8.24s    remaining: 2.05s
900:    learn: 0.2992087        total: 9.22s    remaining: 1.01s
999:    learn: 0.2912932        total: 10.2s    remaining: 0us
> Feature importance:
                     Feature Id  Importances
0                        prefix    19.202524
1                       n_roots     8.639185
2                         depth     7.401090
3                   n_ancestors     7.091980
4          intrinsic_ic_sanchez     4.912018
5         xref__orphanet__count     3.976000
6             xref__omim__count     3.860123
7            xref__mondo__count     3.537002
8             xref__doid__count     3.456866
9                  intrinsic_ic     3.426787
10                is_gwas_trait     3.368879
11  intrinsic_ic_sanchez_scaled     3.051188
12            xref__mesh__count     3.047183
13            xref__ncit__count     2.794949
14                    n_parents     2.678225
15                   n_children     2.488526
16           xref__icd10__count     2.159789
17            xref__umls__count     2.104636
18                n_descendants     2.049742
19            xref__gard__count     2.004128
20        xref__snomedct__count     1.991120
21          xref__meddra__count     1.937532
22          intrinsic_ic_scaled     1.443108
23            xref__icd9__count     1.251675
24          xref__omimps__count     1.146293
25                     n_leaves     0.979452

> Classification report:
                    precision    recall  f1-score   support

01-disease-subtype       0.80      0.85      0.82      1123
   02-disease-root       0.72      0.64      0.68       776
   03-disease-area       0.77      0.79      0.78       242
    04-non-disease       0.97      0.99      0.98       918

          accuracy                           0.83      3059
         macro avg       0.82      0.82      0.82      3059
      weighted avg       0.83      0.83      0.83      3059
```

### CODE REVIEW

File: nxontology_ml/model.py [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+
+
+@dataclass
+class TrainingRecord:
+    efo_otar_slim_id: str
+    efo_label: str
+    rs_classification: str
+
+
+class FeatureType(Enum):
```
Comment (dhimmel) [inline nxontology_ml/model.py:24]: Nice, we often use `StrEnum` / sublcassing `(str, Enum)` for things like this such that the values remain readable. Not a strong preference and if integers are needed then ignore this.

File: nxontology_ml/model.py [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+    @classmethod
+    def from_feature_name(cls, name: str) -> "FeatureType":
+        if name == "__label__":
+            return cls.LABEL
+        if cls._is_numerical(name):
+            return cls.NUMERICAL
+        if cls._is_categorical(name):
+            return cls.CATEGORICAL
+        assert cls._is_ignored(name), f"Unknown feature: {name}"
+        return cls.IGNORED
```
Comment (dhimmel) [inline nxontology_ml/model.py:39]: Great system for now for assigning feature type. In future we can think if it makes sense to define some sort of dataclass or pydantic model where we encode feature type as a field metadata property.

File: nxontology_ml/model.py [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+
+
+def read_training_data(
+    take: int | None = None,
+    data_path: Path = ROOT_DIR / "data/efo_otar_slim_v3.43.0_rs_classification.tsv",
+) -> Iterable[dict[str, Any]]:
+    # Get labelled data
+    def _labelled_data() -> Iterable[TrainingRecord]:
+        data = iter(data_path.read_text().splitlines())
+        assert next(data).strip() == "efo_otar_slim_id	efo_label	rs_classification"
```
Comment (dhimmel) [inline nxontology_ml/model.py:66]: csv module perhaps preferable?

File: nxontology_ml/model.py [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+    nxo = get_efo_otar_slim()
+    nodes: set[str] = set(nxo.graph)
+
+    # Filter nodes from the ontology that are labelled
+    for record in islice(_labelled_data(), take):
+        id_ = record.efo_otar_slim_id
+        if id_ in nodes:
+            metrics = nxo.node_info(id_).get_metrics()
+            metrics["__label__"] = record.rs_classification
+            yield metrics
```
Comment (dhimmel) [inline nxontology_ml/model.py:81]: General commit no action needed.

Note that we can assume that we will never have an ontology where the number of nodes exceeds something that will fit in memory. Therefore, feel liberated to use techniques such as reading into pandas doing a join/merge, etc.

I don't love the pandas API and not insisting we should use it whenever possible, but it's a fair assumption that all collaborators will be familiar with it and therefore it can be a more succinct (sometimes less safe and less disciplined) route to accomplish everything.

File: nxontology_ml/model.py [hunk 5]
```diff
@@ -0,0 +1,152 @@
+from collections.abc import Iterable
+from dataclasses import dataclass
+from enum import Enum
+from itertools import islice
+from pathlib import Path
+from typing import Any
+
+import numpy as np
+import pandas as pd
+from catboost import CatBoostClassifier, FeaturesData, Pool
```
Comment (dhimmel) [inline nxontology_ml/model.py:10]: Curious as to whether catboost is more appropriate than xgboost or lightgbm for our data? Looks nice and simple and I have no clue which is best.

File: nxontology_ml/model.py [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+    train_pool = Pool(data=training_data.train_data, label=training_data.train_labels)
+    test_pool = Pool(data=training_data.test_data, label=training_data.test_labels)
+    model: CatBoostClassifier = CatBoostClassifier(metric_period=100)
+    model.fit(train_pool)
+
+    print("> Feature importance:")
+    print(model.get_feature_importance(test_pool, prettified=True))
+    print()
+    print("> Classification report:")
+    print(classification_report(training_data.test_labels, model.predict(test_pool)))
```
Comment (dhimmel) [inline nxontology_ml/model.py:152]: Are these predictions on samples that the model has been trained on? Is there any sort of cross-validation or regularization to prevent overfitting?

**Review by dhimmel — APPROVED** [review]
Looks great and feature important makes sense to my intuition! Comments mostly questions or general thoughts for the future.

File: nxontology_ml/model.py (hunk 1 above)
Comment (yonromai) [reply nxontology_ml/model.py:24]: Good input, will update!

File: nxontology_ml/model.py (hunk 3 above)
Comment (yonromai) [reply nxontology_ml/model.py:66]: Indeed, will use https://docs.python.org/3/library/csv.html

File: nxontology_ml/model.py (hunk 6 above)
Comment (yonromai) [reply nxontology_ml/model.py:152]: No, by default it's trained on 85% of the data and the evaluation is run on the remaining 15%

File: nxontology_ml/model.py (hunk 5 above)
Comment (yonromai) [reply nxontology_ml/model.py:10]: This choice was arbitrary tbh (I have mostly used xgboost in past professional experiences and wanted to give a try to Catboost). Benchmarking them against each other is something we should do in the near future.

**Comment by yonromai** [comment]
@dhimmel Updated the PR with the csv and StrEnum changes 🙏

File: nxontology_ml/model.py [hunk 7]
```diff
@@ -21,11 +22,11 @@ class TrainingRecord:
...
-class FeatureType(Enum):
-    LABEL = 1
-    NUMERICAL = 2
-    CATEGORICAL = 3
-    IGNORED = 4
+class FeatureType(str, Enum):
+    LABEL = auto()
+    NUMERICAL = auto()
+    CATEGORICAL = auto()
+    IGNORED = auto()
```
Comment (dhimmel) [inline nxontology_ml/model.py:29]: did not know about `auto`, that is cool
